File: util/cpp/net/WakeOnLAN.cpp
/* 
 * WakeOnLAN.cpp
 *
 * This file is part of the IHMC Util Library
 * Copyright (c) 1993-2014 IHMC.
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public License
 * version 3 (GPLv3) as published by the Free Software Foundation.
 *
 * U.S. Government agencies and organizations may redistribute
 * and/or modify this program under terms equivalent to
 * "Government Purpose Rights" as defined by DFARS 
 * 252.227-7014(a)(12) (February 2014).
 *
 * Alternative licenses that allow for use within commercial products may be
 * available. Contact Niranjan Suri at IHMC (nsuri@ihmc.us) for details.
 */

#include "WakeOnLAN.h"

#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#include "Logger.h"

using namespace NOMADSUtil;

#define checkAndLogMsg if (pLogger) pLogger->logMsg

#if defined (UNIX)
    #define stricmp strcasecmp
#endif

WakeOnLAN::WakeOnLAN (void)
{
}

WakeOnLAN::~WakeOnLAN (void)
{
}

int WakeOnLAN::init (void)
{
    int rc;
    if (0 != (rc = _dgSocket.init())) {
        checkAndLogMsg ("WakeOnLAN::init", Logger::L_MildError,
                        "failed to initialize DatagramSocket; rc = %d\n", rc);
        return -1;
    }
    return 0;
}

int WakeOnLAN::wakeUp (const char *pszMACAddr)
{
    int rc;
    uint32 aui32MACAddr [MAC_ADDR_LEN];
    uint8 aui8MACAddr [MAC_ADDR_LEN];
    if (MAC_ADDR_LEN != (rc = sscanf (pszMACAddr, "%2x:%2x:%2x:%2x:%2x:%2x",   /*!!*/ // NOTE: There is a dependency between this format specifier and MAC_ADDR_LEN
                                      &aui32MACAddr[0], &aui32MACAddr[1], &aui32MACAddr[2], &aui32MACAddr[3], &aui32MACAddr[4], &aui32MACAddr[5]))) {
        checkAndLogMsg ("WakeOnLAN::wakeUp1", Logger::L_MildError,
                        "failed to parse MAC address - expecting %d octets; got %d octets\n", MAC_ADDR_LEN, rc);
        return -1;
    }
    for (int i = 0; i < 6; i++) {
        aui8MACAddr[i] = (uint8) aui32MACAddr[i];
    }
    if (0 != (rc = wakeUp (aui8MACAddr))) {
        checkAndLogMsg ("WakeOnLAN::wakeUp1", Logger::L_MildError,
                        "wakeUp2 failed with rc = %d\n", rc);
        return -2;
    }
    return 0;
}

int WakeOnLAN::wakeUp (uint8 *pui8MACAddr)
{
    int rc;
    uint8 aui8WOLPacket [6 + 16 * MAC_ADDR_LEN];
    for (int i = 0; i < 6; i++) {
        aui8WOLPacket[i] = 0xFF;
    }
    for (int i = 0; i < 16 * MAC_ADDR_LEN; i++) {
        aui8WOLPacket[i+6] = pui8MACAddr [i % MAC_ADDR_LEN];
    }
    if ((rc = _dgSocket.sendTo ("255.255.255.255", 7, aui8WOLPacket, sizeof (aui8WOLPacket))) < 0) {
        checkAndLogMsg ("WakeOnLAN::wakeUp2", Logger::L_MildError,
                        "sendTo() on UDPDatagramSocket failed with rc = %d; errno = %d (%s)\n",
                        rc, _dgSocket.getLastError(), strerror (_dgSocket.getLastError()));
        return -1;
    }
    return 0;
}

File: util/java/us/ihmc/util/WakeOnLAN.java
/*
 * WakeOnLAN.java
 *
 * This file is part of the IHMC Util Library
 * Copyright (c) 1993-2016 IHMC.
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public License
 * version 3 (GPLv3) as published by the Free Software Foundation.
 *
 * U.S. Government agencies and organizations may redistribute
 * and/or modify this program under terms equivalent to
 * "Government Purpose Rights" as defined by DFARS 
 * 252.227-7014(a)(12) (February 2014).
 *
 * Alternative licenses that allow for use within commercial products may be
 * available. Contact Niranjan Suri at IHMC (nsuri@ihmc.us) for details.
 */

package us.ihmc.util;

import java.net.InetAddress;
import java.net.DatagramPacket;
import java.net.DatagramSocket;

/**
 * Date: Oct 2, 2009
 * Time: 9:36:33 AM
 * @author  Maggie Breedy <Nomads Team>
 * @version $Revision: 1.6 $
 */

public class WakeOnLAN
{
    public WakeOnLAN()
    {
    }

    public static void main (String[] args) 
    {
        if (args.length != 1) {
            System.out.println("Usage: <MACaddress>");
            System.exit (1);
        }
        String macAddr = args[0];
        WakeOnLAN wakeonLan = new WakeOnLAN();
        wakeonLan.wakeup (macAddr);
    }

    //Makes the wake up call
    public void wakeup (String macAddress)
    {
        try {
            byte[] macByteArray = getMACAddressBytes (macAddress);
            byte[] bytes = new byte [6 + 16 * macByteArray.length];
            for (int i = 0; i < 6; i++) {
                bytes[i] = (byte) 0xff;
            }
            for (int i = 6; i < bytes.length; i += macByteArray.length) {
                System.arraycopy (macByteArray, 0, bytes, i, macByteArray.length);
            }

            System.out.println("Wake-on-LAN: Creating the datagram socket...");

            InetAddress inetAddress = InetAddress.getByName ("255.255.255.255");
            DatagramPacket dgPacket = new DatagramPacket(bytes, bytes.length, inetAddress, PORT);
            DatagramSocket dgSocket = new DatagramSocket();

            System.out.println("Wake-on-LAN: Sending the datagram packet...");
            dgSocket.send (dgPacket);
            dgSocket.close();

            System.out.println ("Wake-on-LAN packet sent.");
        }
        catch (Exception e) {
            System.out.println("Failed to send Wake-on-LAN packet: + e");
            e.printStackTrace();
            System.exit(1);
        }
    }

    //Gets the byte array in hex of the MAC Address 
    protected byte[] getMACAddressBytes (String macAddrStr)
    {
        byte[] byteArray = new byte [6];
        String[] hex = macAddrStr.split ("(\\:|\\-)");

        if (hex.length != 6) {
            System.out.println ("Invalid MAC address.");
        }
        try {
            for (int i = 0; i < 6; i++) {
                byteArray[i] = (byte) Integer.parseInt (hex[i], 16);
            }
        }
        catch (NumberFormatException nfe) {
            nfe.printStackTrace();
        }
        return byteArray;
    }

    //Variables
    public static final int PORT = 5;
}
